### PR TITLE
[Enhancement] reuse lambda argument related sub expressions within a lambda function in projection

### PR DESCRIPTION
**Pull request by fzhedu**
Signed-off-by: fzhedu <fzhedu@gmail.com>

## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

### key idea
lambda functions differ from outer functions in projection,as they have local arguments, they should evaluate with local context. as a result, the common sub expressions should not processed like other functioins in projection's context. Actually, common sub expressions are evaulated within each lambda context, one by one in order. 
as the code, FE extracts common sub expressions within each lambda function, as explain with `[]` follwed lambda functions, in order of slot id. BE evaluates common sub expressions in order.

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->
```
mysql> explain select array_map(x -> x*2 + x*2, [1,3]);
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Explain String                                                                                                                                                                                                             |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                                                                                                                                                            |
|  OUTPUT EXPRS:3: array_map                                                                                                                                                                                                 |
|   PARTITION: UNPARTITIONED                                                                                                                                                                                                 |
|                                                                                                                                                                                                                            |
|   RESULT SINK                                                                                                                                                                                                              |
|                                                                                                                                                                                                                            |
|   1:Project                                                                                                                                                                                                                |
|   |  <slot 3> : array_map(<slot 2> -> <slot 7> + <slot 7>[<slot 5> <-> cast([2, TINYINT, true] as SMALLINT)][<slot 6> <-> [5, SMALLINT, true] * 2][<slot 7> <-> cast([6, SMALLINT, true] as INT)], ARRAY<tinyint(4)>[1,3]) |
|   |                                                                                                                                                                                                                        |
|   0:UNION                                                                                                                                                                                                                  |
|      constant exprs:                                                                                                                                                                                                       |
|          NULL                                                                                                                                                                                                              |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
12 rows in set (0.01 sec)

mysql> select array_map(x -> x*2 + x*2, [1,3]);
+-------------------------------------------+
| array_map(x -> (x * 2) + (x * 2), [1, 3]) |
+-------------------------------------------+
| [4,12]                                    |
+-------------------------------------------+
1 row in set (0.02 sec)
```
## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3
  - [ ] 2.2

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 90 / 106 (84.91%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|3|9|33.33%|[539, 540, 541, 542, 543, 544]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/LambdaFunctionOperator.java|3|5|60.00%|[55, 56]|
|:large_blue_circle:|com/starrocks/sql/ast/LambdaFunctionExpr.java|11|17|64.71%|[43, 44, 72, 79, 80, 88]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|61|63|96.83%|[430, 431]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuseRule.java|12|12|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 92 / 106 (86.79%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|3|9|33.33%|[539, 540, 541, 542, 543, 544]|
|:large_blue_circle:|com/starrocks/sql/ast/LambdaFunctionExpr.java|11|17|64.71%|[43, 44, 72, 79, 80, 88]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|61|63|96.83%|[430, 431]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/LambdaFunctionOperator.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuseRule.java|12|12|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 106 / 107 (99.07%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/LambdaFunctionExpr.java|16|17|94.12%|[88]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|63|63|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/LambdaFunctionOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuseRule.java|12|12|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/ast/LambdaFunctionExpr.java [hunk 1]
```diff
@@ -52,19 +66,45 @@ protected void analyzeImpl(Analyzer analyzer) throws AnalysisException {
...
+        String names = getChild(1).explain();
+        int realChildrenNum = getChildren().size() - 2 * commonSubOperatorNum;
+        if (realChildrenNum > 2) {
+            names = "(" + getChild(1).explain();
+            for (int i = 2; i < realChildrenNum; ++i) {
+                names = names + ", " + getChild(i).explain();
+            }
+            names = names + ")";
+        }
+        StringBuilder commonSubOp = new StringBuilder();
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/ast/LambdaFunctionExpr.java:99]: added

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 118 / 133 (88.72%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/LambdaFunctionExpr.java|18|33|54.55%|[90, 91, 92, 93, 94, 95, 97, 99, 100, 101, 103, 104, 105, 107, 113]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|73|73|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/LambdaFunctionOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuseRule.java|12|12|100.00%|[]|

**Comment by wanpengfei-git** [comment]
@Mergify rebase

**Comment by fzhedu** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 118 / 133 (88.72%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/LambdaFunctionExpr.java|18|33|54.55%|[90, 91, 92, 93, 94, 95, 97, 99, 100, 101, 103, 104, 105, 107, 113]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|73|73|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/LambdaFunctionOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuseRule.java|12|12|100.00%|[]|

**Comment by fzhedu** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java [hunk 2]
```diff
@@ -307,12 +315,115 @@ private boolean isNonDeterministicFuncOrLambdaArgumentExist(ScalarOperator scala
...
                     return true;
                 }
             }
             return false;
         }
 
     }
+
+
+    public static Projection getNewProjection(Projection projection, ColumnRefFactory columnRefFactory,
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java:328]: What is the function of getNewProjection? Is it to calculate common expr? I think the name is a bit too vague, please add a comment and think about a more suitable name?

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/ExpressionTest.java [hunk 3]
```diff
@@ -564,11 +565,43 @@ public void testLambdaReuseSubExpressionWithoutLambdaArguments() throws Exceptio
...
+        sql = "select array_map(x -> x*2 + x*2, [1,3])";
+        plan = getFragmentPlan(sql);
+        Assert.assertTrue(plan.contains("array_map(<slot 2> -> <slot 7> + <slot 7>:common expressions:{<slot 5> <-> " +
+                "CAST(<slot 2> AS SMALLINT)}{<slot 6> <-> <slot 5> * 2}{<slot 7> <-> CAST(<slot 6> AS INT)}, " +
+                "ARRAY<tinyint(4)>[1,3])"));
+
+        sql = "select array_map((x,y) -> x*2 + y  + x*2, [1,3],[4,5])";
+        plan = getFragmentPlan(sql);
+        Assert.assertTrue(plan.contains("array_map((<slot 2>, <slot 3>) -> CAST(CAST(<slot 7> AS INT) + " +
+                "CAST(<slot 3> AS INT) AS BIGINT) + CAST(<slot 7> AS BIGINT):common expressions:{<slot 6> <-> " +
```
Comment (HangyuanLiu) [inline fe/fe-core/src/test/java/com/starrocks/sql/plan/ExpressionTest.java:584]: Could there be a more readable format?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java (hunk 2 above)
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java]: renamed to `rewriteProjectionOrLambdaExpr`

File: fe/fe-core/src/test/java/com/starrocks/sql/plan/ExpressionTest.java (hunk 3 above)
Comment (fzhedu) [reply fe/fe-core/src/test/java/com/starrocks/sql/plan/ExpressionTest.java]: updated

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 120 / 137 (87.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/LambdaFunctionExpr.java|20|37|54.05%|[93, 94, 95, 96, 97, 98, 100, 102, 103, 104, 106, 107, 108, 110, 111, 113, 119]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|73|73|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/LambdaFunctionOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuseRule.java|12|12|100.00%|[]|

**Comment by fzhedu** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 120 / 137 (87.59%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/ast/LambdaFunctionExpr.java|20|37|54.05%|[93, 94, 95, 96, 97, 98, 100, 102, 103, 104, 106, 107, 108, 110, 111, 113, 119]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java|73|73|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/scalar/LambdaFunctionOperator.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/ScalarOperatorToExpr.java|9|9|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuseRule.java|12|12|100.00%|[]|

**Review by HangyuanLiu — APPROVED** [review]

**Comment by fzhedu** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java [hunk 4]
```diff
@@ -269,33 +274,36 @@ private static Set<ScalarOperator> getOperatorsByDepth(int depth,
...
+        // when called this function without reuseLambda, here get rid of the expressions with lambda arguments,
+        // for example, select a+b, array_map(x-> 2*x+2*x > a+b, [1])
         // the a+b is reused, but 2*x is not reused as it contains the lambda argument x.
-        // TODO(fzh) support reusing lambda argument related expressions.
         @Override
-        public Integer visitLambdaFunctionOperator(LambdaFunctionOperator scalarOperator, CollectorContext context) {
-            return collectCommonOperatorsByDepth(scalarOperator.getLambdaExpr().accept(this, new CollectorContext(true)),
-                    scalarOperator, true);
+        public Integer visitLambdaFunctionOperator(LambdaFunctionOperator scalarOperator, Void context) {
+            ignoreLambdaArg = !reuseLambda;
```
[the remark this answers is not in the document]
Comment (fzhedu) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/tree/ScalarOperatorsReuse.java:291]: actually need two, ignoreLambdaArg is true when reuseLambda is false and within a lambda function.

**Comment by fzhedu** [comment]
run starrocks_admit_test

**Review by kangkaisen — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by fzhedu** [comment]
@mergify backport branch-2.5

**Pull request merged**